### DISCUSSION TITLE
What's the advantage of useDispatch over store.dispatch?

### DISCUSSION THREAD

**Discussion by ZYinMD** [reduxjs/react-redux#1789] [Q&A]
Personally, I haven't experienced any problems if I simply use `store.dispatch()` inside my components. 

I even feel it's nicer than `useDispatch` in some cases, where multiple instances of the component are created, e.g. rows in a list or even table cells, because I feel the hook creates an extra "stateful thing" that each instance needs to keep track of individually.

This question has actually been [asked on stackoverflow](https://stackoverflow.com/questions/59339850/what-is-the-advantage-of-usedispatch-over-simply-exporting-dispatch-from-store-c) a long time ago, but didn't draw much interest.

**Comment by markerikson** [comment]
[You should never import the store directly into components](https://redux.js.org/faq/store-setup#can-or-should-i-create-multiple-stores-can-i-import-my-store-directly-and-use-it-in-components-myself).

Importing the store into a component ties that component to that exact store instance. This makes your code harder to test and the component impossible to reuse correctly.  Also, if you end up importing the store just to dispatch actions, but do still use `useSelector` to read data, now you have a mismatch in how your code is interacting with the store.  (and you should _definitely_ never try to subscribe to the store directly in a component.)

_Always_ use the hooks to interact with the store, and wrap the app with `<Provider store={store}>`.

There's no "stateful thing" involved with `useDispatch`.  As you can see here, the hook implementation is trivial - it's literally just returning `store.dispatch`. The important thing is that the store is being dependency-injected via context at runtime:

https://github.com/reduxjs/react-redux/blob/v7.2.4/src/hooks/useDispatch.js#L14-L17

```js
  return function useDispatch() {
    const store = useStore()
    return store.dispatch
  }
```

**Reply by ZYinMD** [reply]
Thanks! One more question about this part:

>  if you end up importing the store just to dispatch actions, but do still use useSelector to read data, now you have a mismatch in how your code is interacting with the store

I never thought it's a problem, what if a component uses useSelector to read data, but can also call some other functions outside of the component, which calls `store.dispatch()`, is that ok? It's literally everywhere in my code, for example:

```js
const foo = {
  go() {
    store.dispatch({ type: "go" });
  },
};

function Bar() {
  const a = useSelector((state) => state.a);
  return <span onClick={() => foo.go()}>{a}</span>;
}

```
(Btw I only have one store.)

**Reply by markerikson** [reply]
It may run, but you're also risking running into circular import problems, and this is _not_ how we want people writing Redux code.

Please don't do this.

**Reply by ZYinMD** [reply]
Thanks! Looking at my code, it seems hard not to use `store.dispatch()`, and it never occurred to me that it is "not how we want people to write Redux". In the following example, how would you suggest a paradigm change?
```js
// CoffeeMachine.tsx
export function CoffeeMachine() {
  const isBrewing = useSelector((state) => state.coffeeMachine.isBrewing);
  if (isBrewing) return <span>Brewing!</span>;
  return <button onClick={() => coffeeMachine.brew()}>Brew!</button>;
}

// coffeeMachine.ts
export const coffeeMachine = {
  waterLevel: 100,
  hasEnoughWater() {
    return this.waterLevel > 50;
  },
  brew() {
    if (!this.hasEnoughWater()) {
      store.dispatch({ type: "coffee_machine_low_in_water" });
      return;
    }
    this.checkPower();
    this.checkPods();
    store.dispatch({ type: "coffee_brewing_starts" });
  },
  checkPower() {
    //...
  },
  checkPods() {
    //...
  },
};

```

**Reply by markerikson** [reply]
That's... a _very_ unusual pattern. I don't think I've _ever_ seen anyone try to organize their Redux logic that way before:

- I've sometimes seen people do `import * as todoActions from './todoSlice'`, and then reference `dispatch(todoActions.todoAdded())`, and I've even seen lots of cases where people tried to put Redux logic in classes, but I've never seen anyone put logic into a plain object like that
- I definitely haven't seen anyone try to use `this` with plain objects and Redux functions like that
- You're keeping state values completely outside of both the Redux store _and_ a React component. Why? that doesn't make any sense that I can think of.

I'm honestly not even sure what that structure is supposed to accomplish.  Uh... for that matter, are you putting functions _into the Redux state_?  **Please don't ever put functions into the Redux state**: https://redux.js.org/style-guide/style-guide#do-not-put-non-serializable-values-in-state-or-actions .

_Normal_ React-Redux usage would be:

- Not importing the store
- Calling `const dispatch = useDispatch()` inside of the component
- Doing `dispatch(someActionCreator())` inside of a click handler
- Writing logic that needs to interact with the store in the form of a thunk ( https://redux.js.org/usage/writing-logic-thunks )

I have to ask, what led you to write the code this way?  We certainly don't show anything like this in our official docs, and I haven't seen anything like this in any of the other tutorials or codebases I've looked at over the years.

**Reply by ZYinMD** [reply]
Hi Mark, thanks very much for your reply.

I'm a little surprised that my pattern is an odd one. The coffeeMachine is a singleton, which represents a coffee machine in the real world. 

Maybe the plain-object approach does look odd, let me copy the exact same code but as a class:
```js
// CoffeeMachine.tsx
export function CoffeeMachine() {
  const isBrewing = useSelector((state) => state.coffeeMachine.isBrewing);
  if (isBrewing) return <span>Brewing!</span>;
  return <button onClick={() => coffeeMachine.brew()}>Brew!</button>;
}

// coffeeMachine.ts
export const coffeeMachine = new (class {
  #waterLevel = 100;
  #hasEnoughWater() {
    return this.#waterLevel > 50;
  }
  public brew() {
    if (!this.#hasEnoughWater()) {
      store.dispatch({ type: "coffee_machine_low_in_water" });
      return;
    }
    this.#checkPower();
    this.#checkPods();
    this.#waterLeve -= 20;
    store.dispatch({ type: "coffee_brewing_starts" });
  }
  #checkPower() {
    //...
  }
  #checkPods() {
    //...
  }
})();

```
A line-by-line comparison would show it's 100% the same as the plain-object before. Sometimes I use plain objects when 1. it's a singleton, 2. it's doesn't need any constructor arguments, 3. it's created at runtime start, not created by some other function calls. Basically it's just like an npm package that you import. (Maybe I'll stop doing that so I don't confuse people.)

But anyway, my point is, in this pattern, the coffeeMachine encapsulates its own logic, states, methods, etc, because it's easy to reason about this way. And it exposes a public method for the react component to call, which is friendly to use. However, some of its methods will inevitably be able to change things on the screen, so these methods call `store.dispatch()`. I wonder if it's bad. I personally haven't experience any issues before, but if today I hear it's bad I'll definitely pay more attention in the future.

============

edit: since you said you "didn't even know what it's supposed to accomplish", I'll add the slice as below, hopefully to make things more clear, but it wasn't a critical part of the example so I didn't include it yesterday:
```js
// coffeeMachineSlice.ts
const coffeeMachineSlice = createSlice({
  name: "coffee_machine",
  initialState: {
    brewing: false,
    waterWarning: false,
  },
  reducers: {
    coffee_brewing_starts: (state) => {
      state.brewing = true;
    },
    coffee_machine_low_in_water: (state) => {
      state.waterWarning = true;
    },
  },
});
```

**Reply by markerikson** [reply]
Yeah, writing a class like that is also definitely _not_ idiomatic React or Redux usage.

In general, the React and Redux ecosystems assume plain JS data and "functional-light" programming style - no classes, no "models", no inheritance, no "public/private".

**Reply by ZYinMD** [reply]
Hi Mark, thanks very much! What's your comment on the following code that uses `store.dispatch()`?

```js
// App.tsx
function App() {
  useEffect(() => {
    onAppStart();
  }, []);
  return (
    <Provider store={store}>
      <div>Hello World</div>
    </Provider>
  );
}

// onAppStart.ts
export function onAppStart() {
  // initGoogleAnalytics();
  // addAllKindsOfEventListeners();
  // applyUserSettingsFromIndexedDB();
  // await login();
  await checkPremiumMembership();
  // startPeriodicSyncing();
}

// checkPremiumMembership.ts
export async function checkPremiumMembership() {
  const status = await fetchPremiumStatus(); /* this could be "in_free_trial" |
    "trial_has_ended" | "has_paid" , etc */
  store.dispatch({ type: "fetched_user_premium_status", payload: status });
}
```
Would you say the `store.dispatch()` here is also not idiomatic? I know it can be avoided by using thunks, but that'll result in dispatching thunks inside thunks inside thunks, and writing curried functions everywhere just to get a hold of `dispatch`, which looks less clean, and doesn't seem to provide much benefit in return.

**Reply by markerikson** [reply]
I'm not sure how much more clearly I can state this:

**Directly importing and using the store is _never_ idiomatic use of Redux.  Please don't do that.**

Thunks _are_ idiomatic usage:

https://redux.js.org/usage/writing-logic-thunks

**Reply by ZYinMD** [reply]
Yes, you have been very clear that `store.dispatch()` is not idiomatic. However, the original question in this thread had always been about "the disadvantage of `store.dispatch()`", and I still fail to see it when I have only one store, except for "not idiomatic".

On the other hand, the disadvantage of thunks are clear in some occasions: just because one tiny function needs to access `dispatch`, its parent functions and grandparent functions ALL need to turn into curried thunks, just for the sake of dispatching the child and child's child. Sometimes it's not a big deal, some other times it's not worth the readability. It'd be ironic if Redux solves "props drilling" but then introduces "dispatch drilling".

**Reply by markerikson** [reply]
I already gave you the answers in my first response: it makes your code harder to test and harder to reuse.
